control/subscription_group.go

```diff
@@ -0,0 +1,339 @@
+	plugins []core.Plugin
+}
+
+//TODO needs godoc
+func (v subscriptionGroups) Add(id string, requested []core.RequestedMetric, metrics []core.Metric,
+	configTree cdata.ConfigDataTree, plugins []core.Plugin) []serror.SnapError {
+	v.Lock()
+	defer v.Unlock()
+	return v.add(id, requested, metrics, configTree, plugins)
+
+}
+
+func (v subscriptionGroups) add(id string, requested []core.RequestedMetric, metrics []core.Metric,
+	configTree cdata.ConfigDataTree, plugins []core.Plugin) []serror.SnapError {
+	if _, ok := v.subscriptionMap[id]; ok {
+		return []serror.SnapError{serror.New(ErrSubscriptionGroupAlreadyExists)}
+	}
+	serrs := v.subscribePlugins(id, plugins)
+	v.subscriptionMap[id] = &subscriptionGroup{
+		requested:  requested,
+		metrics:    metrics,
+		plugins:    plugins,
+		configTree: configTree,
+	}
+	return serrs
+}
+
+//TODO needs godoc
+func (s subscriptionGroups) Remove(id string) []serror.SnapError {
+	s.Lock()
+	defer s.Unlock()
+	return s.remove(id)
+}
+
+func (s subscriptionGroups) remove(id string) []serror.SnapError {
+	if _, ok := s.subscriptionMap[id]; !ok {
+		return []serror.SnapError{serror.New(ErrSubscriptionGroupDoesNotExist)}
+	}
+	serrs := s.unsubscribePlugins(id, s.subscriptionMap[id].plugins)
+	delete(s.subscriptionMap, id)
+	return serrs
+}
+
+//TODO needs godoc
+func (s subscriptionGroups) Get(id string) ([]core.RequestedMetric, []core.Metric,
+	[]core.Plugin, cdata.ConfigDataTree, error) {
+	s.Lock()
+	defer s.Unlock()
+	return s.get(id)
+}
+
+func (s subscriptionGroups) get(id string) ([]core.RequestedMetric, []core.Metric,
+	[]core.Plugin, cdata.ConfigDataTree, error) {
+	if _, ok := s.subscriptionMap[id]; !ok {
+		return nil, nil, nil, cdata.ConfigDataTree{}, ErrSubscriptionGroupDoesNotExist
+	}
+	sg := s.subscriptionMap[id]
+	return sg.requested, sg.metrics, sg.plugins, sg.configTree, nil
+}
+
+func (s subscriptionGroups) update(id string, metrics []core.Metric, plugins []core.Plugin) error {
+	// s.Lock()
+	// defer s.Unlock()
+	if _, ok := s.subscriptionMap[id]; !ok {
+		return ErrSubscriptionGroupDoesNotExist
+	}
+	s.subscriptionMap[id].metrics = metrics
+	s.subscriptionMap[id].plugins = plugins
+	return nil
+}
+
+// Process recalculate the resulting metrics for all subscription groups based on the
+// requested metrics (expanding metrics) also updating the required plugins.  Plugins
+// will be subscribed and unsubscribed based on what is now required by the subscription
+// group.
+func (s *subscriptionGroups) Process() (errs []serror.SnapError) {
+	s.Lock()
+	defer s.Unlock()
+	return s.process()
+}
+
+func (s *subscriptionGroups) process() (errs []serror.SnapError) {
+	for id, group := range s.subscriptionMap {
+		metrics, plugins, serrs := s.getMetricsAndCollectors(group.requested, group.configTree)
+		if serrs != nil {
+			for _, err := range errs {
+				errs = append(errs, err)
+			}
+		}
+
+		//add processors and publishers to collectors just gathered
```

Comment:
@IRCody Do we want to have process/publish plugins always in latest version, don't we? How we could tell that one wants to use processor/publisher plugin in lower version? Correct me if I'm wrong but I think that if one really wants to switch to lower version of processor/publisher, one may load it and unloads the latest version. `process()` will find the only available version (in this case newly loaded lower version) and subscribe it. In case when one loads processor/publisher in latest version, `process()` will also automatically discover it and subscribe it.
When there are multiple version of processor/publisher loaded, always the latest one would be used.